### PR TITLE
DO NOT MERGE: Feat/redataset

### LINKED ISSUE #406: nm_join() (the artist formerly known as redataset() and superset())
Returns a single data frame with model output (from specified table files) and input data.

Similar to the old `metrumrg::superset()` function, resurrected and redesigned by @kylebaron in 2020.

# Proposed code

```r
#' Return a vector of absolute paths to table files to look for
#' 
#' @param mod the `bbi_nonmem_model` object
#' 
#' @details
#' Called by redataset
#' 
tabfiles <- function(mod) {
  output_dir <- get_output_dir(mod)
  c(
    file.path(output_dir, "PAR_TAB"), 
    file.path(output_dir, "TAB"),
    build_path_from_model(mod, "par.tab"), 
    build_path_from_model(mod, ".tab")
    
  )
}


#' Return a single data frame with model output and input data
#' 
#' @param mod the `bbi_nonmem_model` object or a path to a NONMEM run
#' @param join character column name to use to join tab files
#' @param files absolute paths to table files to try to join
#' @param more absolute paths to other files to try to join
#' @param verbose if TRUE, messages are printed as files are read
#' 
#' @details
#' Required packges: data.table, dplyr, bbr. 
#' 
#' Note that you can get the absolute path to the model output directory 
#' with `bbr::get_output_dir(mod)` if needed to build paths for passing to 
#' `files` or `more`.
#' 
redataset <- function(mod, join = "NUM", files = tabfiles(mod), more = NULL, 
                      verbose = TRUE) {
  stopifnot(requireNamespace("data.table"))
  stopifnot(requireNamespace("dplyr"))
  stopifnot(requireNamespace("bbr"))
  if (!inherits(mod, "bbi_nonmem_model")) {
    mod <- read_model(mod)
  }
  files <- c(files,more)
  verbose <- isTRUE(verbose)
  sum <- bbr::model_summary(mod)
  nid <-  sum$run_details$number_of_subjects
  nrec <- sum$run_details$number_of_data_records
  data_loc <- get_data_path(sum)
  if(verbose) message("data file: ", basename(data_loc))
  
  data <- data.table::fread(data_loc, na = '.', data.table = FALSE)
  data <- dplyr::as_tibble(data)
  if(verbose) {
    message("  rows: ", nrow(data))
    message("  cols: ", ncol(data))
  }
  chk <- file.exists(files)
  files <- files[chk]
  if(length(files)==0) {
    message("  zero table files found; returning")
    return(data)
  }
  if(!all(join %in% names(data))) {
    stop("couldn't find '", join, "' column in 'data'")  
  }
  leading_cols <- names(data)
  for(p in files) {
    tab <- data.table::fread(p, skip = 1, data.table = FALSE, na ='.')
    tab <- dplyr::as_tibble(tab)
    has_id <- "ID" %in% names(tab)
    if(!(nrow(tab) %in% c(nrec,nid))) {
      if(verbose) {
        message("\ntable file: ", basename(p), " (skipped)")
        message("  rows: ", nrow(tab))
        message("  hasid: ", has_id)
        message("  tabids: ", length(unique(tab[["ID"]])))
        message("  runids: ", nid)
      }
      next;
    }
    if(nrow(tab) == nid && has_id) {
      if(verbose) {
        message("\nfirstonly file: ", p)
        message("  rows: ", nrow(tab))
        message("  cols: ", ncol(tab))
      }
      keep <- c("ID",setdiff(names(tab), names(data)))
      if((ncol(tab) - length(keep) - 1) > 0) {
        if(verbose) {
          message(
            "  droppping: ", 
            ncol(tab) - length(keep) - 1, 
            " columns in common"
          )
        }
      }
      data <- dplyr::left_join(data, tab[,keep], by = "ID", suffix = c("", "y"))
      next;
    } # end ID join
    if(nrow(tab) != nrec) {
      message("\ntable file: ", basename(p), " (skipped)")
      message("  rows: ", nrow(tab))
      message("  hasid: ", has_id)
      message("  tabids: ", length(unique(tab[["ID"]])))
      message("  runids: ", nid)
      next;
    }
    if(verbose) message("\ntable file: ", basename(p))
    new_cols <- setdiff(names(tab), names(data))
    keep <- c(join,new_cols)
    drop <- setdiff(names(tab), new_cols)
    drop <- drop[!(drop %in% join)]
    if(verbose) {
      message("  rows: ", nrow(tab))
      message("  cols: ", length(new_cols), " new")
      for(d in drop) {
        message("  drop: ", d,  " common")  
      }
    }
    data <- dplyr::left_join(tab[,keep], data, by = join)
  }  
  if(verbose) {
    message("\nfinal stats:")
    message("  rows: ", nrow(data))
    message("  cols: ", ncol(data))
  }
  dplyr::select(data, dplyr::all_of(leading_cols), dplyr::everything())
}
```

### PR DESCRIPTION
**Pull request by seth127**
Moving this functionality to a fresh branch for cleanliness. This turned into a dev/experimentation branch that ended up spawning #426 and #429. 

# Original description

Closes #406 

Manual testing can be done with the following:

```r
# clone the repo and checkout `feat/redataset` branch
# make sure you have bbi installed and `options('bbr.bbi_exe_path')` points to it
devtools::load_all()
setwd("tests/testthat")
source("setup-workflow-ref.R")
?nm_join
df <- nm_join(MOD1)
```

Also check out:

* the [`nm_join() code](https://github.com/metrumresearchgroup/bbr/blob/feat/redataset/R/nm-join.R)
* any of the tests in [`tests/testthat/test-nm-join.R`](https://github.com/metrumresearchgroup/bbr/blob/feat/redataset/tests/testthat/test-nm-join.R)
* the control stream and output files in [`inst/model/nonmem/basic/1`](https://github.com/metrumresearchgroup/bbr/tree/feat/redataset/inst/model/nonmem/basic)

### CODE REVIEW

**Comment by kylebaron** [comment]
If `superset` is `FALSE`, then the first table that is processed needs to be `nrec` rows long; if not, decline to process.

File: R/nm-join.R [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+#'
+#' @importFrom dplyr left_join select all_of everything
+#' @importFrom readr read_csv read_table
+#' @export
+nm_join <- function(
+  .mod,
+  .join_col = "NUM",
+  .files = tabfiles(.mod),
+  .more = NULL,
+  .verbose = getOption("bbr.verbose"),
```
Comment (dpastoor) [inline R/nm-join.R:51]: In the first zoom call code review when we discussed moving this to getOption, I might have been too focused on the parameter still being there and just switching it to an option default. as i was thinking about it - the verbosity here probably doesn't even need an arg anymore, you could just query for verbosity within the function. Idiomatically suppression of chatty messages via such options is common, i'm not sure if anyone would actually use this. Is there a user reason to want to tune this more easily.

I'm thinking of things like dplyr::summarize, tidyverse etc where they don't have args for messages anywhere, its just an option to set.

**Review by dpastoor — COMMENTED** [review]
some misc comments - overall been lightly following convo in slack, good to see this coming together quickly

File: R/nm-join.R (hunk 1 above)
Comment (seth127) [reply R/nm-join.R:53]: hmm, good point. Yea, I'll think on that. There are a couple other functions with some verbosity that I was thinking about switching over to use the new option too. Thanks.

**Comment by seth127** [comment]
`nm_table_files()`

* parses `$TABLE ... FILE=` from control stream and returns (absolute?) file paths
* maybe the `nm_tables()` (and `nm_join()`) either expect absolute paths or paths relative to `get_output_dir(.mod)`

`nm_tables()`

```r
list(
  data = tibble(the input data file),
  tables = list(
    `tab` = tibble(the first table file),
    `par.tab` = tibble(the second table file),
  )
)
```

for the names in the $tables list, use basename(filenames): 

* remove `get_model_id()` from the front and back of file name
* use `make.names`
* remove any leading or trailing punctuation
* check for name collisions before naming

**Pull request closed without merging**